\PassOptionsToPackage{utf8}{inputenc}
%\documentclass{article}

\documentclass[11pt,hidelinks]{article}
%\usepackage[margin=0.75in,top=0.75in,footskip=0.5in]{geometry}
\usepackage{comment}

\usepackage{helvet}
\renewcommand{\familydefault}{\sfdefault}

\usepackage[draft]{hyperref}
\usepackage{authblk}
\usepackage{fullpage}

\newcommand{\vocab}{\textbf}

\begin{document}

\title{Pangenome graph layout by HOGWILD! path-guided stochastic gradient descent}

\author{ %[Heumos, \textit{et~al}.]{
  Simon Heumos\,$^{1 \dagger}$,
  Andrea Guarracino\,$^{2 \dagger}$,
  Pjotr Prins\,$^3$,
  and Erik Garrison\,$^{3 *}$
}

\affil{\small$^1$Quantitative Biology Center (QBiC), University of T\"ubingen, T\"ubingen, Germany, 72076 \\
  $^2$University of Rome ``Tor Vergata'', Rome, Italy \\
  $^3$UTHSC, Memphis, TN, USA \\
  $^\ast$To whom correspondence should be addressed. \\
  $^\dagger$Contributed equally.}

\maketitle

\abstract{High-quality low-cost genome assemblies led to sequencing of whole populations whose genomic variation can be comparatively studied in a graphical pangenome. In pangenome graphs models, DNA sequences are incorporated as nodes with edges connecting the nodes as they occur as sequences representing the graph. Pangenome graphs built from raw sets of alignments may have complex structures generated by common patterns of genome variation. These nonlinear structures can introduce difficulty in downstream analyses, visualization, and interpretation. We propose a new layout algorithm that orders the nodes of a pangenome graph using a HOGWILD! path-guided stochastic gradient descent approach: PG-SGD. Our implementation demonstrates that we can efficiently compute the latent structure of gigabase-scale pangenome graphs revealing their underlying biology.
}
\section{Introduction}

Pangenome graphs \cite{computational2016computational}.

\section{Implementation}

\section{Evaluation}

\section{Discussion}

\section*{Funding}

\bibliography{document}
\bibliographystyle{ieeetr}

\end{document}
