\PassOptionsToPackage{utf8}{inputenc}
\documentclass{bioinfo}

\usepackage{makecell}

\usepackage{floatrow}

\usepackage{comment}

\usepackage{siunitx}

% singlelinecheck=false puts subcaptions on the left
\usepackage[singlelinecheck=false]{subcaption}

\usepackage{amsthm}
\theoremstyle{definition}
\newtheorem{definition}{Definition}[section]
\newtheorem{theorem}{Theorem}[section]
\newtheorem{corollary}{Corollary}[theorem]
\newtheorem{lemma}[theorem]{Lemma}

\usepackage{amsfonts}
\usepackage{booktabs}

\usepackage{algorithm2e}
\usepackage[usenames,dvipsnames]{xcolor}
\SetAlgoLined
\usepackage{bm}

% we squeeze our figures even more together
\captionsetup{belowskip=-2pt}

\SetAlgoLined
\SetKwProg{MyStruct}{Struct}{ contains}{end}

\newcommand{\vocab}{\textbf}
\newcommand{\red}[1]{{\textcolor{Red}{#1}}}
\newcommand{\FIXME}[1]{\red{[FIXME: #1]}}

\usepackage{orcidlink}
\hypersetup{hidelinks}
\usepackage{appendix}
\newcommand{\description}{}
\usepackage[inline]{enumitem}


\def\labelitemi{--}

\copyrightyear{2023} \pubyear{XXXX}

\access{Advance Access Publication Date: Day Month Year}
\appnotes{Genome Analysis}

\begin{document}
    \firstpage{1}

    \subtitle{Genome Analysis}

    %\title[Pangenome graph layout by HOGWILD! Path-Guided Stochastic Gradient Descent]{Pangenome graph layout by HOGWILD! Path-Guided Stochastic Gradient Descent}
    %\title[Pangenome graph layout by Path-Guided Stochastic Gradient Descent]{Pangenome graph layout by Path-Guided Stochastic Gradient Descent}
    %\title[Unveiling genome variation by sequence-guided pangenome graph layouts]{Unveiling genome variation by sequence-guided pangenome graph layouts}
    %\title[PG-SGD: sequence-guided pangenome graph layouts]{PG-SGD: sequence-guided pangenome graph layouts}
    %\title[Genome-guided pangenome graph layouts via Stochastic Gradient Descent]{Genome-guided pangenome graph layouts via Stochastic Gradient Descent}
	%\title[Sequence-guided pangenome graph layouts via Stochastic Gradient Descent]{Sequence-guided pangenome graph layouts via Stochastic Gradient Descent}
    %\title[Genome-guided pangenome graph layouts]{Genome-guided pangenome graph layouts}
	\title[Sequence-guided pangenome graph layouts]{Sequence-guided pangenome graph layouts}
    
	\author[Heumos, Guarracino \textit{et~al}.]{
        Simon~Heumos\,$^{\orcidlink{0000-0003-3326-817X}1,2,\dagger}$,
        Andrea~Guarracino\,$^{\orcidlink{0000-0001-9744-131X}3,4,\dagger}$,
        Jan-Niklas Manuel Schmelzle\,$^{\orcidlink{0000-0001-8566-4049}5,6}$,
        Jiajie Li\,$^{\orcidlink{0000-0002-9467-0659}6}$,
        Zhiru Zhang\,$^{\orcidlink{0000-0002-0778-0308}6}$,
        Sven Nahnsen\,$^{\orcidlink{0000-0002-4375-0691}1,2}$,
        Pjotr Prins\,$^{\orcidlink{0000-0002-8021-9162}3}$,
        Erik~Garrison\,$^{\orcidlink{0000-0003-3821-631X}\text{\sfb 3},*}$
    }

    \address{
        $^1$Quantitative Biology Center (QBiC), University of Tübingen, Tübingen 72076, Germany \\
        $^2$Biomedical Data Science, Department of Computer Science, University of Tübingen, Tübingen 72076, Germany \\
        $^3$Department of Genetics, Genomics and Informatics, University of Tennessee Health Science Center, Memphis, TN 38163, USA \\
        $^4$Genomics Research Centre, Human Technopole, Milan 20157, Italy \\
        $^5$Department of Computer Engineering, School of Computation, Information and Technology (CIT), Technical University of Munich, Munich 80333, Germany \\
        $^6$School of Electrical and Computer Engineering, Cornell University, Ithaca, NY 14853, USA \\
    }

    \corresp{
        $^\ast$To whom correspondence should be addressed. \\
        $^{\dagger}$The authors wish it to be known that, in their opinion, the first two authors should be regarded as Joint First Authors.\
    }

    \history{Received on XXXXX; revised on XXXXX; accepted on XXXXX}

    \editor{Associate Editor: XXXXXXX}

	\abstract{
	\textbf{Motivation:}
	The increasing availability of complete genomes demands models to study genomic variability within entire populations.
	Pangenome graphs can represent the full genetic diversity between multiple genomes, but they may exhibit complex structures due to common, nonlinear patterns of genome variation and evolution.
	These structures introduce difficulty in downstream analyses, visualization, and interpretation. \\
	\textbf{Results:}
	To address these challenges, we propose a novel graph layout algorithm, the Path-Guided Stochastic Gradient Descent (PG-SGD).
	PG-SGD uses the genomes, represented in the pangenome graph as paths, to move pairs of nodes in parallel by applying the HOWGWILD! strategy.
	We show that our implementation efficiently computes the layout of gigabase-scale pangenome graphs, revealing their biological features. \\
	\textbf{Availability:}
	PG-SGD has been integrated in \textit{ODGI} which is released as free software under the MIT open source license.
	Source code is available at \url{https://github.com/pangenome/odgi}. \\% and documentation at \url{https://odgi.readthedocs.io/en/latest/rst/tutorials/sort_layout.html}. \\
	\textbf{Contact:} \href{egarris5@uthsc.edu}{egarris5@uthsc.edu} \\
	\textbf{Supplementary information:} Supplementary data are available at \textit{Bioinformatics} online.
	}
	
	\maketitle
	
	\section{Introduction}
	Reference genomes are the most widely used resource in genetics, serving as a foundation for a variety of analyses
	such as gene annotation, read mapping, and variant detection \citep{Singh2022}.
	However, with the availability of hundreds to thousands of high quality genomes per species, this linear model has become obsolete.
	A single genome can not faithfully represent the genetic diversity of a species, leading to the reference bias problem \citep{Ballouz2019}.
	Rather, a pangenome models the entire set of genomic elements of a given population \citep{Tettelin_2008,cpang2018,Eizenga_2020, Sherman_2020}.
	Pangenomes can be represented by a sequence graph incorporating sequences as nodes and their connections as edges \citep{Hein1989}.
	These nodes are shared for identical sequences, such as homologs, paralogs, and orthologs.
	In the variation graph model \citep{Garrison:2018}, genomes are encoded as paths visiting the nodes in the graph.
	%A bidirected graph can even contain both strands of DNA as paths. \\
	
	A pangenome graph layout is the arrangement of nodes and edges in an \textit{N}-dimensional space such that a human accessible representation of genetic variation across multiple genomes can be drawn. 
	Layout algorithms aim to find suitable node coodinates in order to minimize overlapping nodes or edges, reduce edge crossings, and promote intuitive understanding by displaying the graph layout.
	One popular approach is force-directed graph drawing \citep{cheong_force-directed_2022} which produces aesthetically-pleasing layouts.
	This is prone to get stuck in local minima, but stochastic gradient descent (SGD) implementations can alleviate such a problem \citep{Zheng2019}.
	Typically, force-directed layouts are hard to compute \citep{wang_research_2014}, but the HOGWILD! SGD is a parallelizable, therefore scalable solution \citep{Recht2011}.
	In practice, multidimensional scaling (MDS) is applied to minimize the difference between the visual distance and theoretical graph distance. This can be accomplished by using pairwise node	 distances to minimize an energy function.
	%\FIXME{
	%	Existing layout algorithms range from simple, force-directed methods to more complex, hierarchical layouts, each presenting unique strengths and drawbacks.
	%	Force-directed methods offer an aesthetic layout based on a physics simulation but may struggle with scalability for larger graphs (XXX WE NEED CITATIONS XXX).
	%	In contrast, hierarchical layouts handle layered relationships and dependencies, yet can exhibit limitations when dealing with non-hierarchical data (XXX WE NEED CITATIONS XXX)
	%}
	Since pangenome graphs represent genomes as paths in the graph, a reasonable distance metric would be the nucleotide distance between a pair of nodes traversed by the same path.
	But to our knowledge, no algorithm takes into account such biological information to compute the pangenome graph layouts.
	
	Here, we present a new pangenome graph layout algorithm which applies a path-guided stochastic gradient descent (PG-SGD) to move pairs of nodes in a HOGWILD! parallel manner.
	The algorithm computes the pangenome graph layout that best reflects the nucleotide sequences in the graph.
	It can be extended in any number of dimensions. We provide in the ODGI toolkit \citep{Guarracino2022} implementations to produce layouts in 1 dimension (1D) and 2 dimensions (2D).
	% We may want to adjust some citations using https://tex.stackexchange.com/questions/585942/how-to-display-the-first-two-authors-in-a-citation-call-out.

	\section{Algorithm}
	%\FIXME{WE ONLY WANT TO TALK ABOUT LAYOUTS AND PRODUCING LAYOUTS, NOT ABOUT GRAPH SORTING. PG-SGD PRODUCES 1D AND 2D LAYOUTS. IN 1D WE SORT A GRAPH BY PROJECTING THE ACTUAL 1D LAYOUT INTO A NODE ORDER!}
	
	PG-SGD is inspired by \cite{Zheng2019}, but we designed our algorithm to work on pangenomes in the form of a variation graph introduced in Definition \ref{def:vg}.

	\begin{definition}
		\label{def:vg}
		Variation graphs are a mathematical formalism to represent pangenome graphs \citep{Garrison_2019_thesis}.
		In the variation graph $\mathcal{G} = (\mathcal{V}, \mathcal{E}, \mathcal{P})$, nodes (or vertices) $\mathcal{V} = v_1\ldots v_{|\mathcal{V}|}$ contain nucleotide sequences.
		Each node $v_i$ has a unique identifier $i$ and an implicit reverse complement $\bar{v_i}$.
		The node strand $s$ represents node orientation.
		Edges $\mathcal{E} = e_1\ldots e_{|\mathcal{E}|}$ connect ordered pairs of node strands ($e_i = ( s_a, s_b )$), defining the graph topology.
		Paths $\mathcal{P} = p_1\ldots p_{|\mathcal{P}|}$ describe walks over node strands ($p_i = s_1 \ldots s_{|p_i|}$), representing the genomes embedded in the graph.
	\end{definition}

	In brief, the algorithm moves a single pair of nodes $\mathcal{V}_i$, $\mathcal{V}_j$ at a time, minimizing the disparity between the layout distance $ld_{ij}$ of a node pair and the nucleotide distance $nd_{ij}$ of the same nodes as calculated along a path that traverses them.
	%Our multithreaded implementation (\url{https://odgi.readthedocs.io/en/latest/rst/tutorials/sort_layout.html}) presents a working prototype that is based on succinct graph data structures (\url{https://doi.org/ghqdzw}):
	\begin{enumerate*}[label=(\roman*)]
		\item The first node $\mathcal{V}_i$ of a pair is a uniform path step pick from all nodes.
		\item The second node $\mathcal{V}_j$ of a pair is sampled from the same path following a Zipfian distribution \citep{Zipf1932}.
		\item The path nucleotide distance  $nd_{ij}$ guides the actual layout distance $ld_{ij}$ update of these nodes. Also referred to as \textit{term update}.
		\item The magnitude $r$ of the update depends on the current learning rate $\mu$ of the SGD.
	\end{enumerate*}
	These key steps are sketched in Figure \FIXME{Add puritan figure.}. Pseudocode is shown in Algorithm~\ref{alg:pg_sgd}.
	
	\input{pg-sgd_pseudocode}
	
	The number of iterations steer the annealing step size $\eta$ which in turn determines the learning rate $\mu$ of PG-SGD. 
	Since variation graphs are very sparse, a large step size in the beginning of the schedule already leads to a globally linear \FIXME{(or planar?)} layout.
	With decreasing $\eta$ the node layout adjustments become more local, ensuring that the nodes are positioned in away to best reflect the nucleotde distances of the paths down to the single base pair level. 
	
	In standard SGD layout algorithms, the problem space is given by pairwise node comparisons. In PG-SGD, the problem space is given by the pairwise path step comparisons. The number of path steps in large variation graphs can be several orders of magnitude higher than the number of nodes. Therefore, the number of term updates in PG-SGD is bound to the number of path steps in the graph.
	%Should the delta of $nd_{ij}$ and $ld_{ij}$ become 0 before the annealing schedule is finished, the algorithm stops.

	Originating from empirical inspection of word frequency tables, Zipf's law states that a word with rank $n$ occurs $1/n$ times as the most frequent one.
	It's mathematical formulation is called the Zipfian distribution whichs rank-frequency distribution is related to the inverse power law.
	Sampling from a Zipfian distribution generated with a path's nucleotide position space increases the possibility to draw a small nucleotide position.
	So there is a high chance of a small nucleotide distance between node pair $\mathcal{V}_ij$ to refine the layout down to the nodes containing only one base pair.
	
	To provide balance between global and local term updates, there is a 50\% chance for a \textit{flip}: In half of our term updates the second path step is sampled from a uniform distribution instead of a Zipfian one. Additionally, after half the number of iterations, a \textit{cooling} phase skewes the Zipfian distribution so that the possibility of a a small sampled nucleotide distance is significantly increased. Again, this boosts the local planarity of the graph layout.
	
	\section{Implementation}
	
	We implemented the algorithm in ODGI which serves as an interface to graphs in the GFA (\url{https://github.com/GFA-spec/GFA-spec}) format. 
	Our path index is a strict subset of the XG index~\citep{Garrison:2018} with the exception that
	we specifically avoided to use succinct SDSL~\citep{Gog2014} data structures. We only use bit compressed integer vectors. 
	This enables efficient retrieval of path positions to quickly compute nucleotide distances and eliminates the need to store all pairwise distances between nodes in memory, allowing PG-SGD to be applied to gigabase-scale pangenome graphs.

	Our implementation is multithreaded, threads can perform term update calculations in parallel. Layout values are stored in an atomic integer vector presenting the only multithreading bottleneck. Effectively, this presents a modified version of the HOGWILD! SGD algorithm which just writes the layout updates to a shared non-atomic integer vector. We went for an atomic vector, because this ensures memory saftey and in our experience there was basically no performance loss.
	
	Due to the fact that we have a very high learning rate, our 1D implementation can work with a random layout as input. Though, by default, it just makes the given node order the starting layout. In 2D, we offer several random layout initializations: 
		\begin{enumerate*}[label=(\roman*)]
		\item Uniform,
		\item gaussian,
		\item or Hilbert.
	\end{enumerate*}

    \iffalse
    Fig. 1: Describe how our approach works, especially a single update operation (Fig. \ref{fig:sketches}). Explanation of 1D graph updating in Figures \ref{fig:1d_before_update}-\ref{fig:1d_after_update}. Explanation of 2D graph updating in Figures \ref{fig:2d_before_update}-\ref{fig:2d_after_update}. Zipfian distribution.

    \input{fig_sketches}
    \fi

    \section{Results}
    \label{sec:results}

    \subsection{Performance}
	\FIXME{Describe the dataset (HPRC chromosomes), runtime and memory usage (specifying the fastest/slowest chromosome pangenome graph layout and the thinnest/fattest in memory), just text no with a table}
    \FIXME{A comparison with Bandage might be worth, to give readers an idea of the scalability}
	

    \subsection{Pangenome graphs layouts reveal biology features}
	\FIXME{With the HPRC dataset, Describe a full chromosome, and 1 or a few selected loci, HLA, C4, beta-defensin locus, LPA, stuff from the HPRC paper or PGR-TK.
	In the supp a huge picture with a "karyotype", that is all chromosome layout together sorted by chromosome number, might be hot}

	\iffalse
	Fig. 2: Performance evaluation 1D + 2D: Time + RAM by number of haplotypes (Fig.~\ref{fig:haps_time_ram}). Time + RAM by number of threads (Fig.~\ref{fig:threads_time_ram}). Full HPRC graph.
    \\
    1D (max. threads) vs. ALIBI; we should compare time + RAM in Fig.~\ref{fig:alibi}.
    But also by our sorting goodness metrics: odgi sort + the one suggested in the ALIBI paper in Table 1.
    \\
    2D (max. threads) + odgi draw vs. Bandage in Fig.~\ref{fig:bandage}.
    \input{fig_performance}
    \\
    \\
    Table 1: Metrics of the ALIBI and PG-SGD graphs from Fig. 2.
    \input{tab_performance.tex}
    Already, PG-SGD outperforms existing graph linearization methods like the flow procedure (\url{https://doi.org/gdw58w}) or ALIBI (\url{https://doi.org/hkv3}).

    \paragraph{Latent graph structure reveals underlying biology}
    Fig. 3: Cool quantitative 1D sortings and 2D layouts: biological implications.
    Randomly sorted. PG-SGD sorted. Ygs sorted. Reference sorted.
    We want a pipeline of sortings. 2D layout of the whole HPRC.
    Chr6 HPRC HLA graph? Chr8 beta-defensin gene cluster HPRC? Whole HPRC?
    \\
    We could also try to build a gastric cancer pangenome graph with data from \url{https://www.nature.com/articles/s41467-022-33073-7} from up to 185 samples.
    However, we would have to request access to the data.
    \\
    \\
    \input{fig_latent_graph_structure}
    Table 2: Metrics of the sorted graphs in Fig. 3.
    \input{tab_metrics_latent_graph_structure}

    \paragraph{Bonus Section}
    Fig. 4: Detect tension. Relax a graph. Detect tension afterwards.
    I need to test this on a new data set I got from Erik.
    I need to establish a fixed lower boundary for the tension from which on we don't relax anymore. \\
    With a high quality layout, we can measure the discrepancy of the path layout position versus the expected path nucleotide position, the “tension” of a graph.
    The greater the “tension”, the greater is the possibility of a biologically meaningless alignment.
    This allows us to detect telomere collapsing alignment errors and hopefully (pangenome) assembly errors, subsequently correcting them.
    \input{fig_tension}
    \fi


	\section{Discussion}
	\label{sec:discussion}
	
	We presented Path-Guided Stochastic Gradient Descent (PG-SGD), the first layout algorithm for pangenome graphs that leverages the biological information available within the genomes represented in the graph.
	Our implementation efficiently computes the layout of pangenome graphs representing thousands of individuals at whole-genome scale.
	\FIXME{Difference to other existing methods, are there possible improvements of our method possible?}
	\FIXME{Discuss performance eval. @Jiajie + Niklas: What about going GPU?}
	
	Graph visualization is key for understanding genome variations and the large-scale layouts produced by PG-SGD offer an unprecedented high-level perspective on variation in pangenomes.
	We implemented PG-SGD to generate layouts in 1 and 2 dimensions.
	These graph projections have already been employed in constructing and analyzing the first draft human pangenome reference \citep{Liao2023}, as well as in the discovery of heterologous recombination of acrocentric chromosomes \citep{Guarracino2023}.
	Furthermore, they are applied in the creation and analysis of pangenome graphs for any species \citep{Guarracino2022, Garrison2023}.
	%The graph simplification pipeline smoothxg runs POA for each block of paths that are co-linear within each seqwish induced variation graph.
	%A prerequisite is that the graph nodes are sorted according to their occurrence in the graph's embedded paths.
	%Our 1D path-guided SGD algorithm is designed to provide this kind of sort.
	%Problem space: We take a look at all paths (FULL HPRC GRAPH ~8 BILLION STEPS ACROSS ~34 THOUSAND PATHS) and not at all nodes (FULL HPRC GRAPH ~100 MILLION). So we have a very large problem space compared to just looking at the nodes.
	There still remains a gap in interactive and scalable solutions that merge layouts of large pangenome graph with annotation.
	Our algorithm will be the foundation of new pangenome graph browsers for exploring meaningful graph layouts (\href{https://github.com/chfi/waragraph}{https://github.com/chfi/waragraph}).
	
	PG-SGD can be extended to any number of dimensions.
	Projecting a graph in a low number of dimensions, but still preserving its biologically relevant information, enable the application of algorithms that use the graph layout for variant detection and classification. 
	Helpful could be to integrate more layout objectives \citep{ahmed_multicriteria_2021} like neighborhood preservation.
	
	Our future research involves leveraging these graph projections to detect structural variants and detect and solve assembly errors.
	Moreover, we consider the extension of the algorithm to RNA and protein sequences.
	
	\section*{Acknowledgments}
	
	We are grateful to members of the HGSVC and HPRC production teams for their development of resources used in our exposition.
	We thank the authors of the pangenome resources made available on GenBank which have made our experiments possible.
	A special thanks goes to Matthias Seybold from the Quantitative Biology Center for maintaining the Core Facility Cluster.
	
	\section*{Funding}
	
	S.H. acknowledges funding from the Central Innovation Programme (ZIM) for SMEs of the Federal Ministry for Economic Affairs and Energy of Germany.
	S.N. acknowledges Germany’s Excellence Strategy (CMFI), EXC-2124 and (iFIT)—EXC 2180–390900677.
	This work was supported by the BMBF-funded de.NBI Cloud within the German Network for Bioinformatics Infrastructure (de.NBI) (031A532B, 031A533A, 031A533B, 031A534A, 031A535A, 031A537A, 031A537B, 031A537C, 031A537D and 031A538A).
	A.G. acknowledges efforts by Nicole Soranzo to establish a pangenome research unit at the Human Technopole in Milan, Italy.
	JNM.S., J.L., and Z.Z. acknowledge funding from the NSF PPoSS Award \#2118709.
	
	\section*{Competing interests}
	The authors declare that they have no competing interests.
	
	\section*{Data availability}
	
	Code and links to data resources used to build this manuscript and its figures can be found in the paper's public repository: \url{https://github.com/pangenome/sorting-paper}.
	
	\section{Animations}
	\FIXME{ADD LINKS TO OUR ANIMATIONS ON ZENODO}
	
	\bibliographystyle{natbib}
	
	\bibliography{document}
	
	\begin{appendices}
	    \section{Appendix}
	    %\FIXME{@ANDREA: Please add some more details about your clever quantization. Why. How. Where. When. Secs.} \\
	    %\FIXME{PGR-TK Paper - Layouts of challenging regions of the human pangenome. - https://www.nature.com/articles/s41592-023-01914-y} \\
	    \FIXME{From 1D layout to 1D node order} \\
	    \FIXME{reference-guided path-guided stochastic gradient descent}
	\end{appendices}

\end{document}
