\PassOptionsToPackage{utf8}{inputenc}
%\documentclass{article}

\documentclass[11pt,hidelinks]{article}
%\usepackage[margin=0.75in,top=0.75in,footskip=0.5in]{geometry}
\usepackage{comment}

\usepackage{helvet}
\renewcommand{\familydefault}{\sfdefault}

\usepackage[draft]{hyperref}
\usepackage{authblk}
\usepackage{fullpage}
% singlelinecheck=false puts subcaptions on the left
\usepackage[singlelinecheck=false]{subcaption}
%\usepackage{tikz,graphics,color,fullpage,float,epsf,caption,subcaption}
\usepackage{graphicx}

\usepackage{amsfonts}
\usepackage{booktabs}
\usepackage{siunitx}

\usepackage{algorithm2e}
\SetAlgoLined
\SetKwProg{MyStruct}{For each}{:}{end}

\newcommand{\vocab}{\textbf}

\begin{document}

\title{Pangenome graph layout by HOGWILD! path-guided stochastic gradient descent}

\author{ %[Heumos, \textit{et~al}.]{
  Simon Heumos\,$^{1,2 \dagger}$,
  Andrea Guarracino\,$^{3 \dagger}$,
  Jan-Niklas Manuel Schmelzle\,$^{4,5}$,
  Jiajie Li\,$^{5}$,
  Zhiru Zhang\,$^{5}$,
  Sven Nahnsen\,$^{1,2}$,
  Pjotr Prins\,$^6$,
  and Erik Garrison\,$^{6 *}$
}

\affil{\small$^1$Quantitative Biology Center (QBiC), University of Tübingen, Tübingen 72076, Germany \\
  $^2$Biomedical Data Science, Department of Computer Science, University of Tübingen , Tübingen 72076, Germany \\
  $^3$Genomics Research Centre, Human Technopole, Milan 20157, Italy \\
  $^4$Department of Computer Engineering, School of Computation, Information and Technology (CIT), Technical University of Munich, Munich 80333, Germany \\
  $^5$School of Electrical and Computer Engineering, Cornell University, Ithaca, NY 14853, USA \\
  $^6$Department of Genetics, Genomics and Informatics, University of Tennessee Health Science Center, Memphis, TN 38163, USA \\
  $^\ast$To whom correspondence should be addressed. \\
  $^\dagger$Contributed equally.}

\maketitle

\abstract{
    High-quality low-cost genome assemblies led to sequencing of whole populations whose genomic variation can be comparatively studied in a graphical pangenome.
    In pangenome graphs models, DNA sequences are incorporated as nodes with edges connecting the nodes as they occur as sequences representing the graph.
    Pangenome graphs built from raw sets of alignments may have complex structures generated by common patterns of genome variation.
    These nonlinear structures can introduce difficulty in downstream analyses, visualization, and interpretation.
    We propose a new layout algorithm that orders the nodes of a pangenome graph using a HOGWILD! path-guided stochastic gradient descent approach: PG-SGD.
    Our implementation demonstrates that we can efficiently compute the latent structure of gigabase-scale pangenome graphs revealing their underlying biology.
}

\section{Background}

Pangenome graphs \cite{computational2016computational}. blubb.
\\
variation graph model
\\
SGD-layout. Layout in general.
\\

\section{Results}

\paragraph{The HOGHWILD! path-guided stochastic gradient descent algorithm}
Our algorithm is inspired by Zheng et al. (\url{https://doi.org/hkv4}) and it is applicable in 1D and in 2D.
In our implementation, the algorithm moves a single pair of nodes at a time, minimizing the disparity between the layout distance of a node pair and the actual nucleotide distance of a path traversing these nodes.
Our path index enables efficient navigation of path positions by updating nodes on the fly, instead of keeping all pairwise distances of all nodes in memory.
This allows the PG-SGD to be applied to gigabase-scale pangenome graphs. \\
Our multi-threaded implementation (\url{https://odgi.readthedocs.io/en/latest/rst/tutorials/sort_layout.html}) presents a working prototype that is based on succinct graph data structures (\url{https://doi.org/ghqdzw}).
The exploration of the path-guided SGD parameter space is in progress, in order to get the best layout as quickly as possible.
Already, the PG-SGD outperforms existing graph linearization methods like the flow procedure (\url{https://doi.org/gdw58w}), or ALIBI (\url{https://doi.org/hkv3}).
\\
We describe in algorithm~\ref{alg:pg_sgd} how the PG-SGD works on a single thread.
\begin{algorithm}
	\MyStruct{Iteration}{
		Perform term update.
	}
	\caption{TODO: PG-SGD algorithm.}
	\label{alg:pg_sgd}
\end{algorithm}

Fig. 1: Describe how our approach works (Fig. \ref{fig:algorithm}). Explanation of 1D graph sorting and metric with Simon's handmade figures \ref{fig:sorting_explained},\ref{fig:metrics_explained}. SGD figure \ref{fig:pg_sgd_update_sketch}. Zipfian distribution figure \ref{fig:fig:zipf_curve}.

\input{fig_algorithm}

\paragraph{Performance evaluation}
Fig. 2: Performance evaluation 1D + 2D: Time + RAM by number of haplotypes (Fig.~\ref{fig:haps_time_ram}). Time + RAM by number of threads (Fig.~\ref{fig:threads_time_ram}). Chr6 HPRC graph?
\\
1D (max. threads) vs. ALIBI; we should compare time + RAM in Fig.~\ref{fig:alibi}.
But also by our sorting goodness metrics: odgi sort + the one suggested in the ALIBI paper in Table 1.
\\
2D (max. threads) + odgi draw vs. Bandage in Fig.~\ref{fig:bandage}.
\input{fig_performance}
\\
\\
Table 1: Metrics of the ALIBI and PG-SGD graphs from Fig. 2.
\input{tab_performance.tex}
\paragraph{Latent graph structure reveals underlying biology}
Fig. 3: Cool quantitative 1D sortings and 2D layouts: biological implications.
andomly sorted. PG-SGD sorted. Ygs sorted. Reference sorted.
We want a pipeline of sortings. 2D layout of the whole HPRC.
Chr6 HPRC HLA graph? Chr8 beta-defensin gene cluster HPRC? Whole HPRC?
\\
We could also try to build a gastric cancer pangenome graph with data from \url{https://www.nature.com/articles/s41467-022-33073-7} from up to 185 samples.
However, we would have to request access to the data.
\\
\\
\input{fig_latent_graph_structure}
Table 2: Metrics of the sorted graphs in Fig. 3.
\input{tab_metrics_latent_graph_structure}
\paragraph{Bonus Section}
Fig. 4: Detect tension. Relax a graph. Detect tension afterwards.
I need to test this on a new data set I got from Erik.
I need to establish a fixed lower boundary for the tension from which on we don't relax anymore. \\
With a high quality layout, we can measure the discrepancy of the path layout position versus the expected path nucleotide position, the “tension” of a graph.
The greater the “tension”, the greater is the possibility of a biologically meaningless alignment.
This allows us to detect telomere collapsing alignment errors and hopefully (pangenome) assembly errors, subsequently correcting them.
\input{fig_tension}

\section{Discussion}

We propose / implemented ....
\paragraph{}
Difference to other existing methods, are there possible improvements of our method possible?
\paragraph{}
Discuss performance eval
@Jiajie + Niklas: What about going GPU?
\paragraph{}
The algorithm allows to inspect a pangenome graph on base-level, as a whole.
\paragraph{from the ODGI paper:}
Its static, large-scale 1D and 2D visualizations of the pangenome graphs allow an unprecedented high-level perspective on variation in pangenomes, and have also been critical in the development of pangenome graph building methods.
However, an interactive solution that combines the 1D and 2D layout of a graph with annotation and read mapping information across different zoom levels is still missing.
Recent interactive pangenome graph browsers are reference-centric (Beyer et al., 2019; Yokoyama et al., 2019), have a limited predefined coordinate system (Durant et al., 2021), or focus primarily on 2D representations (Gonnella et al., 2019; Wick et al., 2015).
Our graph sorting and layout algorithms can provide the foundation for future tools of this type.
We plan to focus on using these learned models to detect structural variation and assembly errors.
\paragraph{}
The graph simplification pipeline smoothxg runs POA for each block of paths that are co-linear within each seqwish induced variation graph.
A prerequisite is that the graph nodes are sorted according to their occurrence in the graph's embedded paths.
Our 1D path-guided SGD algorithm is designed to provide this kind of sort.
Already, the 1D PG-SGD is a key step in the PanGenome Graph Building (PGGB) pipeline that we successfully applied to build the first draft human pangenome reference (Liao et al., bioRxiv 2022).
\paragraph{}
What about the future?
\paragraph{}
How can other scientists benefit from this work?
\section{Conclusion}

\section{Methods}

\bibliography{document}
\bibliographystyle{ieeetr}

\section*{Acknowledgements}
The authors thank members of the HPRC Pangenome Working Group for their insightful discussion and feedback, and members of the HPRC production teams for their development of resources used in our exposition.

\section*{Funding}
JNM.S., J.L., and Z.Z. acknowledge funding from the NSF PPoSS Award \#2118709.
S.H. acknowledges funding from the Central Innovation Programme (ZIM) for SMEs of the Federal Ministry for Economic Affairs and Energy of Germany.
S.N. acknowledges Germany’s Excellence Strategy (CMFI), EXC-2124 and (iFIT)—EXC 2180–390900677.
This work was supported by the BMBF-funded de.
NBI Cloud within the German Network for Bioinformatics Infrastructure (de.NBI) [031A537B, 031A533A, 031A538A, 031A533B, 031A535A, 031A537C, 031A534A and 031A532B].
A.G. acknowledges Nicole Soranzo to establish a pangenome research unit at the Human Technopole in Milan, Italy.

\section*{Competing interests}
The authors declare that they have no competing interests.

\section*{Availability of data and materials}

\end{document}
