\PassOptionsToPackage{utf8}{inputenc}
\documentclass{bioinfo}

\usepackage{makecell}

\usepackage{floatrow}

\usepackage{comment}

\usepackage{siunitx}

% singlelinecheck=false puts subcaptions on the left
\usepackage[singlelinecheck=false]{subcaption}

\usepackage{amsthm}
\theoremstyle{definition}
\newtheorem{definition}{Definition}[section]
\newtheorem{theorem}{Theorem}[section]
\newtheorem{corollary}{Corollary}[theorem]
\newtheorem{lemma}[theorem]{Lemma}

\usepackage{amsfonts}
\usepackage{booktabs}

\usepackage{algorithm2e}
\usepackage[usenames,dvipsnames]{xcolor}
\SetAlgoLined
\usepackage{bm}

% we squeeze our figures even more together
\captionsetup{belowskip=-2pt}

\SetAlgoLined
\SetKwProg{MyStruct}{Struct}{ contains}{end}

\newcommand{\vocab}{\textbf}
\newcommand{\red}[1]{{\textcolor{Red}{#1}}}
\newcommand{\FIXME}[1]{\red{[FIXME: #1]}}

\usepackage{orcidlink}
\hypersetup{hidelinks}
\usepackage{appendix}


\def\labelitemi{--}

\copyrightyear{2023} \pubyear{XXXX}

\access{Advance Access Publication Date: Day Month Year}
\appnotes{Genome Analysis}

\begin{document}
    \firstpage{1}

    \subtitle{Genome Analysis}

    \title[Pangenome graph layout by HOGWILD! Path-Guided Stochastic Gradient Descent]{Pangenome graph layout by HOGWILD! Path-Guided Stochastic Gradient Descent}
    \author[Heumos, Guarracino \textit{et~al}.]{
        Simon~Heumos\,$^{\orcidlink{0000-0003-3326-817X}1,2,\dagger}$,
        Andrea~Guarracino\,$^{\orcidlink{0000-0001-9744-131X}3,4,\dagger}$,
        Jan-Niklas Manuel Schmelzle\,$^{\orcidlink{0000-0001-8566-4049}5,6}$,
        Jiajie Li\,$^{\orcidlink{0000-0002-9467-0659}6}$,
        Zhiru Zhang\,$^{\orcidlink{0000-0002-0778-0308}6}$,
        Sven Nahnsen\,$^{\orcidlink{0000-0002-4375-0691}1,2}$,
        Pjotr Prins\,$^{\orcidlink{0000-0002-8021-9162}3}$,
        Erik~Garrison\,$^{\orcidlink{0000-0003-3821-631X}\text{\sfb 3},*}$
    }

    \address{
        $^1$Quantitative Biology Center (QBiC), University of Tübingen, Tübingen 72076, Germany \\
        $^2$Biomedical Data Science, Department of Computer Science, University of Tübingen, Tübingen 72076, Germany \\
        $^3$Department of Genetics, Genomics and Informatics, University of Tennessee Health Science Center, Memphis, TN 38163, USA \\
        $^4$Genomics Research Centre, Human Technopole, Milan 20157, Italy \\
        $^5$Department of Computer Engineering, School of Computation, Information and Technology (CIT), Technical University of Munich, Munich 80333, Germany \\
        $^6$School of Electrical and Computer Engineering, Cornell University, Ithaca, NY 14853, USA \\
    }

    \corresp{
        $^\ast$To whom correspondence should be addressed. \\
        $^{\dagger}$The authors wish it to be known that, in their opinion, the first two authors should be regarded as Joint First Authors.\
    }

    \history{Received on XXXXX; revised on XXXXX; accepted on XXXXX}

    \editor{Associate Editor: XXXXXXX}

	\abstract{
	\textbf{Motivation:}
	The increasing availability of high quality genomes demands models to study genomic variability within entire populations.
	Pangenome graphs can represent the full genetic diversity between multiple genomes, but they may exhibit complex structures due to common, nonlinear patterns of genome variation and evolution.
	These structures introduce difficulty in downstream analyses, visualization, and interpretation. \\
	\textbf{Results:}
	To address these challenges, we propose a novel graph layout algorithm, the Path-Guided Stochastic Gradient Descent (PG-SGD).
	Our HOWGWILD! method uses the genomes, represented in the pangenome graph as paths, to move pairs of nodes in parallel.
	We show that our implementation efficiently computes the layout of gigabase-scale pangenome graphs, revealing their biological features. \\
	\textbf{Availability:}
	The PG-SGD algorithm has been integrated in \textit{ODGI} which is released as free software under the MIT open source license.
	Source code is available at \url{https://github.com/pangenome/odgi}. \\ % and documentation at \url{https://odgi.readthedocs.io/en/latest/rst/tutorials/sort_layout.html}. \\
	\textbf{Contact:} \href{egarris5@uthsc.edu}{egarris5@uthsc.edu} \\
	\textbf{Supplementary information:} Supplementary data are available at \textit{Bioinformatics} online.
	}
	
	\maketitle
	
	\section{Introduction}
	Reference genomes are the most widely used resource in genetics, serving as a foundation for a variety of analyses
	such as gene annotation, read mapping, and variant detection \citep{Singh2022}.
	However, with the availability of hundreds to thousands of high quality genomes per species, this linear model has become obsolete.
	A single genome can not faithfully represent the genetic diversity of a species, leading to the reference bias problem \citep{Ballouz2019}.
	Rather, a pangenome models the entire set of genomic elements of a given population \citep{Tettelin_2008,cpang2018,Eizenga_2020, Sherman_2020}.
	Pangenomes can be represented by a sequence graph incorporating sequences as nodes and their connections as edges \citep{Hein1989}.
	These nodes are shared for identical sequences, such as homologs, paralogs, and orthologs.
	In the variation graph model \citep{Garrison:2018}, genomes are encoded as paths visiting the nodes in the graph.
	%A bidirected graph can even contain both strands of DNA as paths. \\
	
	A pangenome graph layout is the arrangement of nodes and edges in an \textit{N}-dimensional space such that a human accessible representation of genetic variation across multiple genomes can be drawn. 
	Layout algorithms aim to find suitable node coodinates in order to minimize overlapping nodes or edges, reduce edge crossings, and promote intuitive understanding by displaying the graph layout.
	One popular approach is force-directed graph drawing \citep{cheong_force-directed_2022} which produces aesthetically-pleasing layouts.
	This is prone to get stuck in local minima, but stochastic gradient descent (SGD) implementations can alleviate such a problem \citep{Zheng2019}.
	Typically, force-directed layouts are hard to compute \citep{wang_research_2014}, but the HOGWILD! SGD is a parallelizable, therefore scalable solution \citep{Recht2011}.
	In practice, multidimensional scaling (MDS) is applied to minimize the difference between the visual distance and theoretical graphed distance, and this is accomplished by using pairwise nodes distances to minimize an energy function.
	%\FIXME{
	%	Existing layout algorithms range from simple, force-directed methods to more complex, hierarchical layouts, each presenting unique strengths and drawbacks.
	%	Force-directed methods offer an aesthetic layout based on a physics simulation but may struggle with scalability for larger graphs (XXX WE NEED CITATIONS XXX).
	%	In contrast, hierarchical layouts handle layered relationships and dependencies, yet can exhibit limitations when dealing with non-hierarchical data (XXX WE NEED CITATIONS XXX)
	%}
	Since pangenome graphs represent genomes as paths in the graph, a reasonable distance metric would be the nucleotide distance between a pair of nodes traversed by the same path.
	But to our knowledge, no algorithm takes into account such biological information to compute the pangenome graph layouts.
	
	Here, we present a new pangenome graph layout algorithm which applies a path-guided stochastic gradient descent (PG-SGD) to move pairs of nodes in a HOGWILD! parallel manner.
	The algorithm computes the pangenome graph layout that best reflects the nucleotide sequences in the graph.
	Our implementation produces layouts in 1 dimension (1D) and 2 dimensions (2D), but it can be extended in any number of dimensions.
	
	% We may want to adjust some citations using https://tex.stackexchange.com/questions/585942/how-to-display-the-first-two-authors-in-a-citation-call-out.
	

	\section{Algorithm}
	\FIXME{WE ONLY WANT TO TALK ABOUT LAYOUTS AND PRODUCING LAYOUTS, NOT ABOUT GRAPH SORTING. PG-SGD PRODUCES 1D AND 2D LAYOUTS. IN 1D WE SORT A GRAPH BY PROJECTING THE ACTUAL 1D LAYOUT INTO A NODE ORDER!}
	
	In this work, we describe the results in 1 dimension (1D) and 2 dimensions (2D). In this section we will describe. Then we will. And then. \\
	
	\FIXME{Do we need a formal definition of a variation graph like in the seqwish paper?}
	
	Our HOGHWILD! path-guided stochastic gradient descent algorithm is inspired by \cite{Zheng2019} and is theoretically applicable in any number of dimensions. 
	We implemented the algorithm in ODGI \citep{Guarracino2022}, making use of its already proven interface to pangenome graphs in the standard graphical fragment assebmly (GFA) format (\url{https://github.com/GFA-spec/GFA-spec}).
	In our implementation, the algorithm moves a single pair of nodes at a time, minimizing the disparity between the layout distance of a node pair and the actual nucleotide distance of a path traversing these nodes.
	Our path index enables efficient navigation of path positions by updating nodes on the fly, instead of keeping all pairwise distances of all nodes in memory.
	This allows the PG-SGD to be applied to gigabase-scale pangenome graphs. \\
	Our multithreaded implementation (\url{https://odgi.readthedocs.io/en/latest/rst/tutorials/sort_layout.html}) presents a working prototype that is based on succinct graph data structures (\url{https://doi.org/ghqdzw}):
	\begin{itemize}
	    \item The first node Xi of a pair is a uniform path step pick from all nodes.
	    \item The second node Xj of a pair is sampled from the same path following a Zipfian distribution.
	    \item The path nucleotide distance of the nodes in the pair guides the actual layout distance dij update of these nodes.
	    \item The magnitude r of the update depends on the current learning rate of the SGD.
	\end{itemize}
	%The exploration of the path-guided SGD parameter space is in progress, in order to get the best layout as quickly as possible.
	We describe in algorithm~\ref{alg:pg_sgd} how the PG-SGD works on a single thread.
	We can have several worker threads but only one checker thread!
	
	\subsection{general algorithm}
	\subsection{path index}
	\subsection{parallelization in detail}
	\subsection{randomization}
	ZIPF!
	\subsection{annealing step-size $\eta$}
	\subsection{convergence / stop criteria, iterations, term updates}
	\subsection{cooling || flip}
	\subsection{from layout to 1D node order}
	
	\input{pg-sgd_pseudocode}

    \iffalse
    Fig. 1: Describe how our approach works, especially a single update operation (Fig. \ref{fig:sketches}). Explanation of 1D graph updating in Figures \ref{fig:1d_before_update}-\ref{fig:1d_after_update}. Explanation of 2D graph updating in Figures \ref{fig:2d_before_update}-\ref{fig:2d_after_update}. Zipfian distribution.

    \input{fig_sketches}
    \fi

    \iffalse


    \section{Results}
    \label{sec:results}

%\subsection{Experimental insights}

%\subsection{Limitations}

% something for the supplementary

    \paragraph{Performance evaluation}
    Fig. 2: Performance evaluation 1D + 2D: Time + RAM by number of haplotypes (Fig.~\ref{fig:haps_time_ram}). Time + RAM by number of threads (Fig.~\ref{fig:threads_time_ram}). Full HPRC graph.
    \\
    1D (max. threads) vs. ALIBI; we should compare time + RAM in Fig.~\ref{fig:alibi}.
    But also by our sorting goodness metrics: odgi sort + the one suggested in the ALIBI paper in Table 1.
    \\
    2D (max. threads) + odgi draw vs. Bandage in Fig.~\ref{fig:bandage}.
    \input{fig_performance}
    \\
    \\
    Table 1: Metrics of the ALIBI and PG-SGD graphs from Fig. 2.
    \input{tab_performance.tex}
    Already, the PG-SGD outperforms existing graph linearization methods like the flow procedure (\url{https://doi.org/gdw58w}) or ALIBI (\url{https://doi.org/hkv3}).

    \paragraph{Latent graph structure reveals underlying biology}
    Fig. 3: Cool quantitative 1D sortings and 2D layouts: biological implications.
    Randomly sorted. PG-SGD sorted. Ygs sorted. Reference sorted.
    We want a pipeline of sortings. 2D layout of the whole HPRC.
    Chr6 HPRC HLA graph? Chr8 beta-defensin gene cluster HPRC? Whole HPRC?
    \\
    We could also try to build a gastric cancer pangenome graph with data from \url{https://www.nature.com/articles/s41467-022-33073-7} from up to 185 samples.
    However, we would have to request access to the data.
    \\
    \\
    \input{fig_latent_graph_structure}
    Table 2: Metrics of the sorted graphs in Fig. 3.
    \input{tab_metrics_latent_graph_structure}

    \paragraph{Bonus Section}
    Fig. 4: Detect tension. Relax a graph. Detect tension afterwards.
    I need to test this on a new data set I got from Erik.
    I need to establish a fixed lower boundary for the tension from which on we don't relax anymore. \\
    With a high quality layout, we can measure the discrepancy of the path layout position versus the expected path nucleotide position, the “tension” of a graph.
    The greater the “tension”, the greater is the possibility of a biologically meaningless alignment.
    This allows us to detect telomere collapsing alignment errors and hopefully (pangenome) assembly errors, subsequently correcting them.
    \input{fig_tension}
    \fi


	\section{Discussion}
	\label{sec:discussion}
	
	We have presented the PG-SGD layout algorithm for pangenome graphs that leverages the inherent biological information available within the graph's represented genomic sequences.
	Our implementation efficiently computes the layout of pangenome graphs representing thousands of individuals at whole-genome scale.
	\FIXME{Difference to other existing methods, are there possible improvements of our method possible?}
	\FIXME{Discuss performance eval. @Jiajie + Niklas: What about going GPU?}
	We implemented the PG-SGD algorithm to generate layouts in 1D and 2D, but the algorithm can in princible be extended to any number of dimensions.
	Graph visualization is key for understanding genome variations and the large-scale layouts produced by the PG-SGD algorithm offer an unprecedented 
	high-level perspective on variation in pangenomes. These layouts are critical for pangenome graph building (\FIXME{CITE PGG and HPRC MAIN PAPER}) and 
	making new biological discoveries (\FIXME{CITE ACROCENTRIC PAPER}).
	
	Problem space: We take a look at all paths (FULL HPRC GRAPH ~8 BILLION STEPS ACROSS ~34 THOUSAND PATHS) and not at all nodes (FULL HPRC GRAPH ~100 MILLION). So we have a very large problem space compared to just looking at the nodes.
	%The graph simplification pipeline smoothxg runs POA for each block of paths that are co-linear within each seqwish induced variation graph.
	%A prerequisite is that the graph nodes are sorted according to their occurrence in the graph's embedded paths.
	%Our 1D path-guided SGD algorithm is designed to provide this kind of sort.
	%Already, the 1D PG-SGD is a key step in the PanGenome Graph Building (PGGB) pipeline that we successfully applied to build the first draft human pangenome reference (Liao et al., bioRxiv 2022).
	
	However, there remains a gap in interactive and scalable solutions that merge layouts of large pangenome graph with annotation.
	Our PG-SGD algorithm will be the foundation of new pangenome graph browsers for exploring meaningful graph layouts.
	\FIXME{2D: waragraph + gfaestus for interactive visualization?}
	Futhermore, projecting a graph in a low number of dimensions, but still preserving its biologically relevant information, enable the application of algorithms that use the graph layout to look at variant classification.
	Our future research involve exploiting these learned projections to detect structural variations and assembly errors.
	\FIXME{PROTEIN?}
	
	Already, the 1D PG-SGD is a key step in pangenome graph construction \citep{Liao2023, Garrison2023} and analysis \citep{Guarracino2022, Guarracino2023}.
	\FIXME{2D: waragraph + gfaestus for interactive visualization? Not sure how to add this here. I am afraid the readers might think we also implemented these tools. Better in the discussion?}. \\
	
	
	\section*{Acknowledgments}
	
	We are grateful to members of the HGSVC and HPRC production teams for their development of resources used in our exposition.
	We thank the authors of the pangenome resources made available on GenBank which have made our experiments possible.
	A special thanks goes to Matthias Seybold from the Quantitative Biology Center for maintaining the Core Facility Cluster.
	
	\section*{Funding}
	
	S.H. acknowledges funding from the Central Innovation Programme (ZIM) for SMEs of the Federal Ministry for Economic Affairs and Energy of Germany.
	S.N. acknowledges Germany’s Excellence Strategy (CMFI), EXC-2124 and (iFIT)—EXC 2180–390900677.
	This work was supported by the BMBF-funded de.NBI Cloud within the German Network for Bioinformatics Infrastructure (de.NBI) (031A532B, 031A533A, 031A533B, 031A534A, 031A535A, 031A537A, 031A537B, 031A537C, 031A537D and 031A538A).
	A.G. acknowledges efforts by Nicole Soranzo to establish a pangenome research unit at the Human Technopole in Milan, Italy.
	JNM.S., J.L., and Z.Z. acknowledge funding from the NSF PPoSS Award \#2118709.
	
	\section*{Competing interests}
	The authors declare that they have no competing interests.
	
	\section*{Data availability}
	
	Code and links to data resources used to build this manuscript and its figures can be found in the paper's public repository: \url{https://github.com/pangenome/sorting-paper}.
	
	\bibliographystyle{natbib}
	
	\bibliography{document}
	
	\begin{appendices}
	    \section{Quantization}
	    \FIXME{@ANDREA: Please add some more details about your clever quantization. Why. How. Where. When. Secs.} \\
	    \FIXME{PGR-TK Paper - Layouts of challenging regions of the human pangenome. - https://www.nature.com/articles/s41592-023-01914-y}
	\end{appendices}

\end{document}
