\PassOptionsToPackage{utf8}{inputenc}
\documentclass{bioinfo}

\usepackage{makecell}

\usepackage{floatrow}

\usepackage{comment}

\usepackage{siunitx}

% singlelinecheck=false puts subcaptions on the left
\usepackage[singlelinecheck=false]{subcaption}

\usepackage{amsthm}
\theoremstyle{definition}
\newtheorem{definition}{Definition}[section]
\newtheorem{theorem}{Theorem}[section]
\newtheorem{corollary}{Corollary}[theorem]
\newtheorem{lemma}[theorem]{Lemma}

\usepackage{amsfonts}
\usepackage{booktabs}

\usepackage{algorithm2e}
\usepackage[usenames,dvipsnames]{xcolor}
\SetAlgoLined
\usepackage{bm}

% we squeeze our figures even more together
\captionsetup{belowskip=-2pt}

\SetAlgoLined
\SetKwProg{MyStruct}{Struct}{ contains}{end}

\newcommand{\vocab}{\textbf}
\newcommand{\red}[1]{{\textcolor{Red}{#1}}}
\newcommand{\FIXME}[1]{\red{[FIXME: #1]}}

\usepackage{orcidlink}
\hypersetup{hidelinks}
\usepackage{appendix}
\newcommand{\description}{}
\usepackage[inline]{enumitem}

% table stuff
\usepackage{amsfonts}
\usepackage{booktabs}
\usepackage{siunitx}
\newcommand{\ra}[1]{\renewcommand{\arraystretch}{#1}}
\usepackage{float}
\floatstyle{plaintop}
\restylefloat{table}


\def\labelitemi{--}

\copyrightyear{2023} \pubyear{XXXX}

\access{Advance Access Publication Date: Day Month Year}
\appnotes{Genome Analysis}

\begin{document}
    \firstpage{1}

    \subtitle{Genome Analysis}

    %\title[Pangenome graph layout by HOGWILD! Path-Guided Stochastic Gradient Descent]{Pangenome graph layout by HOGWILD! Path-Guided Stochastic Gradient Descent}
    \title[Pangenome graph layout by Path-Guided Stochastic Gradient Descent]{Pangenome graph layout by Path-Guided Stochastic Gradient Descent}
    %\title[Unveiling genome variation by sequence-guided pangenome graph layouts]{Unveiling genome variation by sequence-guided pangenome graph layouts}
    %\title[PG-SGD: sequence-guided pangenome graph layouts]{PG-SGD: sequence-guided pangenome graph layouts}
    %\title[Genome-guided pangenome graph layouts via Stochastic Gradient Descent]{Genome-guided pangenome graph layouts via Stochastic Gradient Descent}
	%\title[Sequence-guided pangenome graph layouts via Stochastic Gradient Descent]{Sequence-guided pangenome graph layouts via Stochastic Gradient Descent}
    %\title[Genome-guided pangenome graph layouts]{Genome-guided pangenome graph layouts}
	%\title[Sequence-guided pangenome graph layouts]{Sequence-guided pangenome graph layouts}
    
	\author[Heumos, Guarracino \textit{et~al}.]{
        Simon~Heumos\,$^{\orcidlink{0000-0003-3326-817X}1,2,\dagger}$,
        Andrea~Guarracino\,$^{\orcidlink{0000-0001-9744-131X}3,4,\dagger}$,
        Jan-Niklas Manuel Schmelzle\,$^{\orcidlink{0000-0001-8566-4049}5,6}$,
        Jiajie Li\,$^{\orcidlink{0000-0002-9467-0659}6}$,
        Zhiru Zhang\,$^{\orcidlink{0000-0002-0778-0308}6}$,
        Sven Nahnsen\,$^{\orcidlink{0000-0002-4375-0691}1,2}$,
        Pjotr Prins\,$^{\orcidlink{0000-0002-8021-9162}3}$,
        Erik~Garrison\,$^{\orcidlink{0000-0003-3821-631X}\text{\sfb 3},*}$
    }

    \address{
        $^1$Quantitative Biology Center (QBiC), University of Tübingen, Tübingen 72076, Germany \\
        $^2$Biomedical Data Science, Department of Computer Science, University of Tübingen, Tübingen 72076, Germany \\
        $^3$Department of Genetics, Genomics and Informatics, University of Tennessee Health Science Center, Memphis, TN 38163, USA \\
        $^4$Genomics Research Centre, Human Technopole, Milan 20157, Italy \\
        $^5$Department of Computer Engineering, School of Computation, Information and Technology (CIT), Technical University of Munich, Munich 80333, Germany \\
        $^6$School of Electrical and Computer Engineering, Cornell University, Ithaca, NY 14853, USA \\
    }

    \corresp{
        $^\ast$To whom correspondence should be addressed. \\
        $^{\dagger}$The authors wish it to be known that, in their opinion, the first two authors should be regarded as Joint First Authors.\
    }

    \history{Received on XXXXX; revised on XXXXX; accepted on XXXXX}

    \editor{Associate Editor: XXXXXXX}

	\abstract{
	\textbf{Motivation:}
	The increasing availability of complete genomes demands models to study genomic variability within entire populations.
	Pangenome graphs can represent the full genetic diversity between multiple genomes, but their layouts may exhibit complex structures due to common, nonlinear patterns of genome variation and evolution.
	These structures introduce difficulty in downstream analyses, visualization, and interpretation. \\
	\textbf{Results:}
	To address these challenges, we propose a novel graph layout algorithm, the Path-Guided Stochastic Gradient Descent (PG-SGD).
	PG-SGD uses the genomes, represented in the pangenome graph as paths, to move pairs of nodes in parallel by applying a modified HOWGWILD! strategy.
	We show that our implementation efficiently computes the layout of gigabase-scale pangenome graphs, revealing their biological features. \\
	\textbf{Availability:}
	PG-SGD has been integrated in \textit{ODGI} which is released as free software under the MIT open source license.
	Source code is available at \url{https://github.com/pangenome/odgi}. \\% and documentation at \url{https://odgi.readthedocs.io/en/latest/rst/tutorials/sort_layout.html}. \\
	\textbf{Contact:} \href{egarris5@uthsc.edu}{egarris5@uthsc.edu} \\
	\textbf{Supplementary information:} Supplementary data are available at \textit{Bioinformatics} online.
	}
	
	\maketitle
	
	\section{Introduction}
	Reference genomes are the most widely used resource in genetics, serving as a foundation for a variety of analyses
	such as gene annotation, read mapping, and variant detection \citep{Singh2022}.
	However, with the availability of hundreds to thousands of high quality genomes per species, this linear model has become obsolete.
	A single genome can not faithfully represent the genetic diversity of a species, leading to the reference bias problem \citep{Ballouz2019}.
	Rather, a pangenome models the entire set of genomic elements of a given population \citep{Tettelin_2008,cpang2018,Eizenga_2020, Sherman_2020}.
	Pangenomes can be represented by a sequence graph incorporating sequences as nodes and their connections as edges \citep{Hein1989}.
	These nodes are shared for identical sequences, such as homologs, paralogs, and orthologs.
	In the variation graph model \citep{Garrison:2018}, genomes are encoded as paths visiting the nodes in the graph.
	%A bidirected graph can even contain both strands of DNA as paths. \\
	
	A pangenome graph layout is the arrangement of nodes and edges in an \textit{N}-dimensional space such that a human accessible representation of genetic variation across multiple genomes can be drawn. 
	Layout algorithms aim to find suitable node coodinates in order to minimize overlapping nodes or edges, reduce edge crossings, and promote intuitive understanding by displaying the graph layout.
	One popular approach is force-directed graph drawing \citep{cheong_force-directed_2022} which produces aesthetically-pleasing layouts.
	This is prone to get stuck in local minima, but stochastic gradient descent (SGD) implementations can alleviate such a problem \citep{Zheng2019}.
	Typically, force-directed layouts are hard to compute \citep{wang_research_2014}, but the HOGWILD! SGD is a highly parallelizable, therefore scalable solution that can be applied when the optimization problem is sparse \citep{Recht2011}.
	In practice, multidimensional scaling (MDS) is applied to minimize the difference between the visual distance and theoretical graph distance.
	This can be accomplished by using pairwise node distances to minimize an energy function.
	%\FIXME{
	%	Existing layout algorithms range from simple, force-directed methods to more complex, hierarchical layouts, each presenting unique strengths and drawbacks.
	%	Force-directed methods offer an aesthetic layout based on a physics simulation but may struggle with scalability for larger graphs (XXX WE NEED CITATIONS XXX).
	%	In contrast, hierarchical layouts handle layered relationships and dependencies, yet can exhibit limitations when dealing with non-hierarchical data (XXX WE NEED CITATIONS XXX)
	%}
	Since pangenome graphs represent genomes as paths in the graph, a reasonable distance metric would be the nucleotide distance between a pair of nodes traversed by the same path.
	To our knowledge, no algorithm takes into account such biological information to compute the pangenome graph layouts.
	
	Here, we present a new pangenome graph layout algorithm which applies a path-guided stochastic gradient descent (PG-SGD) to move pairs of nodes in parallel with a modified HOGWILD! strategy.
	The algorithm computes the pangenome graph layout that best reflects the nucleotide sequences in the graph.
	PG-SGD can be extended in any number of dimensions. We provide in the ODGI toolkit \citep{Guarracino2022} implementations to produce layouts in 1 dimension (1D) and 2 dimensions (2D).
	% We may want to adjust some citations using https://tex.stackexchange.com/questions/585942/how-to-display-the-first-two-authors-in-a-citation-call-out.

	\section{Algorithm}
	%\FIXME{WE ONLY WANT TO TALK ABOUT LAYOUTS AND PRODUCING LAYOUTS, NOT ABOUT GRAPH SORTING. PG-SGD PRODUCES 1D AND 2D LAYOUTS. IN 1D WE SORT A GRAPH BY PROJECTING THE ACTUAL 1D LAYOUT INTO A NODE ORDER!}
	
	PG-SGD is inspired by \cite{Zheng2019}, but we designed the algorithm to work on pangenomes in the form of a variation graph (Definition \ref{def:vg}).
	
	\begin{definition}
		\label{def:vg}
		Variation graphs are a mathematical formalism to represent pangenome graphs \citep{Garrison_2019_thesis}.
		In the variation graph $\mathcal{G} = (\mathcal{V}, \mathcal{E}, \mathcal{P})$, nodes (or vertices) $\mathcal{V} = v_1\ldots v_{|\mathcal{V}|}$ contain nucleotide sequences.
		Each node $v_i$ has a unique identifier $i$ and an implicit reverse complement $\bar{v_i}$.
		The node strand $o$ represents node orientation.
		Edges $\mathcal{E} = e_1\ldots e_{|\mathcal{E}|}$ connect ordered pairs of node strands ($e_i = ( o_a, o_b )$), defining the graph topology.
		Paths $\mathcal{P} = p_1\ldots p_{|\mathcal{P}|}$ are series of connected steps $s_i$ that refer to node strands in the graph ($p_i = s_1 \ldots s_{|p_i|}$); the paths represent the genomes embedded in the graph.
	\end{definition}

	PG-SGD's pseudocode is shown in Algorithm~\ref{alg:pg_sgd} and key steps are sketched in Figure \FIXME{Add puritan figure.}.
	In brief, the algorithm moves one pair of nodes $( v_i, v_j )$ at a time, minimizing the difference between the layout distance $ld_{ij}$ of the two nodes and the nucleotide distance $nd_{ij}$ of the same nodes as calculated along a path that traverses them.
	%Our multithreaded implementation (\url{https://odgi.readthedocs.io/en/latest/rst/tutorials/sort_layout.html}) presents a working prototype that is based on succinct graph data structures (\url{https://doi.org/ghqdzw}):
	$v_i$ is the node associated with the step $s_i$ sampled uniformly from all the steps in $\mathcal{P}$.
	$v_j$ is the node associated with the step $s_j$ sampled from the same path of $s_i$ by following a Zipfian distribution \citep{Zipf1932}.
	The difference between $nd_{ij}$ and $ld_{ij}$ guides the update of the node coordinates in the layout.
	The magnitude $r$ of the update depends on the learning rate $\mu$ of the SGD.
	The number of iterations steers the annealing step size $\eta$ which determines the learning rate $\mu$.
	%A key property of pangenome graph layouts is that their nodes can be arranged so that most edges go between nodes that are close together in the node order \citep{Guarracino2022}.
	A large $\eta$ in the first iterations leads to a globally linear (in 1D) or planar (in 2D) layout.
	By decreasing $\eta$, the layout adjustments become more localized, ensuring that the nodes are positioned in a way to best reflect the nucleotide distances in the paths (i.e., in the genomes). % LIAR: down to the single base pair level.
	
	\input{pg-sgd_pseudocode}

	%In standard SGD layout algorithms, the problem space is given by pairwise node comparisons. In PG-SGD, the problem space is given by the pairwise path step comparisons. The number of path steps in large variation graphs can be several orders of magnitude higher than the number of nodes. Therefore, the number of term updates in PG-SGD is bound to the number of path steps in the graph.
	%Should the delta of $nd_{ij}$ and $ld_{ij}$ become 0 before the annealing schedule is finished, the algorithm stops.

	Originating from empirical inspection of word frequency tables, Zipf's law states that a word with rank $n$ occurs $1/n$ times as the most frequent one.
	It's mathematical formulation is called the Zipfian distribution and its rank-frequency distribution is related to the inverse power law.
	Sampling $s_j$ from a Zipfian distribution fixed in the $s_i$'s path position space increases the possibility to draw a nucleotide position close to $s_i$.
	So there is a high chance to use small nucleotide distances $nd_{ij}$ to refine the layout of nodes comprising a few base pairs.

	To provide balance between global and local layout updates, in half of the updates ($flip$ flag in Algorithm~\ref{alg:pg_sgd}) the $s_j$ is sampled uniformly instead from a Zipfian distribution (uniform sampling favours global updates).
	Additionally, to boost the local linearity (in 1D) or planarity (in 2D) of the graph layout, a \textit{cooling} phase skewes the Zipfian distribution after half the number of iterations, so that the possibility of a a small sampled nucleotide distance is significantly increased. 

	\section{Implementation}
	
	We implemented PG-SGD in ODGI \citep{Guarracino2022}. % which serves as an interface to graphs in the GFA (\url{https://github.com/GFA-spec/GFA-spec}) format. 
	Our path index is a strict subset of the XG index~\citep{Garrison:2018} where we specifically avoid using succinct SDSL data structures \citep{Gog2014}.
	Instead, we only use bit compressed integer vectors, enabling efficient retrieval of path positions to quickly compute nucleotide distances without the need to store all pairwise distances between nodes in memory.
	This allows PG-SGD to scale on large pangenome graphs representing thousands of whole genomes.

	Graph layout initialization can significantly influence the quality of the final layout.
	In the 1D implementation, nodes are placed in the same order as they appear in the input graph by default, although support for random layout initialization is also provided.
	This random placement ensures that each run of the PG-SGD method could potentially yield a different result, thereby increasing the likelihood of obtaining a meaningful layout.
	In 2D, we offer several layout initialization techniques.
	One option is to position nodes in the first layout dimension according to their order in the input graph and to add either uniform or Gaussian noise in the second dimension.
	Another possibility is to arrange nodes along a Hilbert curve, which often favors the creation of planar final layouts.

	Our implementation is multithreaded and uses shared memory for the layout vector following the HOGWILD! strategy \citep{Recht2011}: Threads perform layout updates without any locking for additional speed up.
	This is possible because pangenome graphs are typically sparse \citep{Guarracino2022}, with low average node degree, meaning most updates only modify small parts of the entire layout.
	The HOGWILD! SGD algorithm writes the layout updates to a shared non-atomic double vector, whereas PG-SGD stores node coordinates in a vector of atomic doubles.
	Such a vector ensures that no memory overwrites are possible. Our tests revealed basically no performance loss with respect to the non-atomic counterpart.
	
    \iffalse
    Fig. 1: Describe how our approach works, especially a single update operation (Fig. \ref{fig:sketches}). Explanation of 1D graph updating in Figures \ref{fig:1d_before_update}-\ref{fig:1d_after_update}. Explanation of 2D graph updating in Figures \ref{fig:2d_before_update}-\ref{fig:2d_after_update}. Zipfian distribution.

    \input{fig_sketches}
    \fi

    \section{Results}
    \label{sec:results}

    \subsection{Performance}
	We apply PG-SGD to the recent draft human pangenome reference \citep{Liao2023} from the Human Pangenome Reference Consortium (HPRC) to show the scalability of the algorithm with a real pangenome.
	Experiments were conducted on a cluster with 24 \texttt{Regular} nodes (32 cores / 64 threads with two AMD EPYC 7343 processors with 512 GB RAM) and 4 \texttt{HighMem} nodes (64 cores / 128 threads with two AMD EPYC 7513 processors with 2048 GB RAM).
	We downloaded pangenome graphs for each autosome (24 in total) and for the mitochondrial DNA.
	Each graph represents 90 whole human haplotypes: 44 diploid individuals plus the GRCh38 \citep{Schneider2017} and CHM13 \citep{Nurk_2021} haploid human references.
	See Supplementary Table~\ref{tab:layout} for graph statistics.
	When applied to these pangenome graphs using one \texttt{Regular} node for each calculation, PG-SGD obtains the graph layouts in 50 minutes on average, with the highest run time observed being chromosome 16 (Supplementary Table~\ref{tab:layout}).
	This is expected since chromosome 16 presents one of the highest levels of segmentally duplicated sequence among the human autosomes \citep{Martin2004}.
	Repetitive sequences lead to graph nodes with a very high number of path traversals, which are computationally expensive to work with \citep{Guarracino2022}.
	Memory consumption is 29.66 GB of RAM on average, with the memory peak again occurring with chromosome 16, presumably due to the memory expensive path index building phase.
	Given its scalability, we even applied PG-SGD to the full graph with all chromosomes together using a \texttt{HighMem} node (Supplementary Table~\ref{tab:layout}). BandageNG\footnote{\url{https://github.com/asl/BandageNG} (last accessed Jul 2023)}, the current state-of-the-art for graph visualization, was not able to produce a layout within 7 days, hitting the wall clock time limit of the cluster.
	Overall, PG-SGD is $\approx$8X faster on average than BandageNG while using $\approx$2X less memory.
	% avg time bng 388 min, avg gb ng 66.84

    \subsection{Pangenome graphs layouts reveal biology features}
	Graph visualization is essential for understanding pangenome graphs and the genome variation they represent.
	We show how PG-SGD allows us quickly gain insight into biological data by looking at the graph layout structure.
	\FIXME{Describe a full chromosome and/or 1/a few selected loci, HLA, C4, beta-defensin locus, LPA, stuff from the human paper or PGR-TK.}

	% \iffalse
	% Fig. 2: Performance evaluation 1D + 2D: Time + RAM by number of haplotypes (Fig.~\ref{fig:haps_time_ram}). Time + RAM by number of threads (Fig.~\ref{fig:threads_time_ram}). Full human graph.
    % \\X
    % 1D (max. threads) vs. ALIBI; we should compare time + RAM in Fig.~\ref{fig:alibi}.
    % But also by our sorting goodness metrics: odgi sort + the one suggested in the ALIBI paper in Table 1.
    % \\
    % 2D (max. threads) + odgi draw vs. Bandage in Fig.~\ref{fig:bandage}.
    % \input{fig_performance}
    % \\
    % \\
    % Table 1: Metrics of the ALIBI and PG-SGD graphs from Fig. 2.
    % \input{tab_performance.tex}
    % Already, PG-SGD outperforms existing graph linearization methods like the flow procedure (\url{https://doi.org/gdw58w}) or ALIBI (\url{https://doi.org/hkv3}).
    % \paragraph{Latent graph structure reveals underlying biology}
    % Fig. 3: Cool quantitative 1D sortings and 2D layouts: biological implications.
    % Randomly sorted. PG-SGD sorted. Ygs sorted. Reference sorted.
    % We want a pipeline of sortings. 2D layout of the whole HPRC.
    % Chr6 HPRC HLA graph? Chr8 beta-defensin gene cluster HPRC? Whole HPRC?
    % \\
    % We could also try to build a gastric cancer pangenome graph with data from \url{https://www.nature.com/articles/s41467-022-33073-7} from up to 185 samples.
    % However, we would have to request access to the data.
    % \\
    % \\
    % \input{fig_latent_graph_structure}
    % Table 2: Metrics of the sorted graphs in Fig. 3.
    % \input{tab_metrics_latent_graph_structure}
    % \paragraph{Bonus Section}
    % Fig. 4: Detect tension. Relax a graph. Detect tension afterwards.
    % I need to test this on a new data set I got from Erik.
    % I need to establish a fixed lower boundary for the tension from which on we don't relax anymore. \\
    % With a high quality layout, we can measure the discrepancy of the path layout position versus the expected path nucleotide position, the “tension” of a graph.
    % The greater the “tension”, the greater is the possibility of a biologically meaningless alignment.
    % This allows us to detect telomere collapsing alignment errors and hopefully (pangenome) assembly errors, subsequently correcting them.
    % \input{fig_tension}
    % \fi

	\section{Discussion}
	\label{sec:discussion}
	
	We presented Path-Guided Stochastic Gradient Descent (PG-SGD), the first layout algorithm for pangenome graphs that leverages the biological information available within the genomes represented in the graph.
	Our implementation efficiently computes the layout of pangenome graphs representing thousands of whole genomes.
	
	Graph visualization is key for understanding genome variations and the large-scale layouts produced by PG-SGD offer an unprecedented high-level perspective on variation in pangenomes.
	We implemented PG-SGD to generate layouts in 1 and 2 dimensions.
	These graph projections have already been employed in constructing and analyzing the first draft human pangenome reference \citep{Liao2023}, as well as in the discovery of heterologous recombination of human acrocentric chromosomes \citep{Guarracino2023}.
	Furthermore, they are applied in the creation and analysis of pangenome graphs for any species \citep{Guarracino2022, Garrison2023}.
	%The graph simplification pipeline smoothxg runs POA for each block of paths that are co-linear within each seqwish induced variation graph.
	%A prerequisite is that the graph nodes are sorted according to their occurrence in the graph's embedded paths.
	%Our 1D path-guided SGD algorithm is designed to provide this kind of sort.
	Of note, there still remains a gap in interactive and scalable solutions that merge layouts of large pangenome graph with annotation.
	Our algorithm will be the foundation of new pangenome graph browsers for studying graph layouts and the genome variation they represent\footnote{\url{https://github.com/chfi/waragraph} (last accessed Jul 2023)}.
	
	\FIXME{@Jiajie + Niklas: Maybe address our SMT issue with your cache locality as a potential solution. Additionally just 2/3 sentences about going GPU?}

	PG-SGD can be extended to any number of dimensions.
	It can be seen as a graph embedding algorithm that converts high-dimensional, sparse pangenome graphs into low-dimensional, dense, and continuous vector spaces, while preserving its biologically relevant information.
	This enables the application of machine learning algorithms that use the graph layout for variant detection and classification. 
	Our future research involves leveraging these graph projections to detect structural variants and to identify and correct assembly errors
	Moreover, we are considering extending the algorithm to RNA and protein sequences to support pantranscriptome graphs \citep{sibbesen_haplotype-aware_2023} and panproteome graphs \citep{dabbaghie_panpa:_2023}, respectively.
	%Helpful could be to integrate more layout objectives \citep{ahmed_multicriteria_2021} like neighborhood preservation.

	\section*{Acknowledgments}
	
	%We are grateful to members of the HGSVC and HPRC production teams for their development of resources used in our exposition.
	%We thank the authors of the pangenome resources made available on GenBank which have made our experiments possible.
	We thank Matthias Seybold from the Quantitative Biology Center for maintaining the Core Facility Cluster.
	
	\section*{Funding}
	
	S.H. acknowledges funding from the Central Innovation Programme (ZIM) for SMEs of the Federal Ministry for Economic Affairs and Energy of Germany.
	S.N. acknowledges Germany’s Excellence Strategy (CMFI), EXC-2124 and (iFIT)—EXC 2180–390900677.
	This work was supported by the BMBF-funded de.NBI Cloud within the German Network for Bioinformatics Infrastructure (de.NBI) (031A532B, 031A533A, 031A533B, 031A534A, 031A535A, 031A537A, 031A537B, 031A537C, 031A537D and 031A538A).
	A.G. acknowledges efforts by Nicole Soranzo to establish a pangenome research unit at the Human Technopole in Milan, Italy.
	JNM.S., J.L., Z.Z., P.P., and E.G. acknowledge funding from the NSF PPoSS Award \#2118709.
	
	\section*{Competing interests}
	The authors declare that they have no competing interests.
	
	\section*{Data availability}
	
	Code and links to data resources used to build this manuscript and its figures can be found in the paper's public repository: \url{https://github.com/pangenome/sorting-paper}.
		
	\bibliographystyle{natbib}
	
	\bibliography{document}
	
	\begin{appendices}
	    \section{Supplementary data}
	    %\FIXME{@ANDREA: Please add some more details about your clever quantization. Why. How. Where. When. Secs.} \\
	    %\FIXME{PGR-TK Paper - Layouts of challenging regions of the human pangenome. - https://www.nature.com/articles/s41592-023-01914-y} \\
		\FIXME{ADD LINKS TO OUR ANIMATIONS ON ZENODO} \\
	    \FIXME{From 1D layout to 1D node order} \\
	    \FIXME{reference-guided path-guided stochastic gradient descent}
	    \subsection{Performance evaluation}
	    \label{sec:performance}
	    \input{table_layout}
		\subsection{Graph layouts}
		\label{sec:layouts}
		\input{fig_2d_layouts}
		\FIXME{1D layouts?}
	\end{appendices}

\end{document}
